scipy/integrate/_quadrature.py

```diff
@@ -1035,3 +1038,235 @@ def newton_cotes(rn, equal=0):
+    draw `n_offsets` i.i.d. random samples from a population of integral
+    estimates. The sample mean :math:`m` of these integral estimates is an
+    unbiased estimator of the true value of the integral, and the standard
+    error of the mean :math:`s` of these estimates may be used to generate
+    confidence intervals using the t distribution with ``n_offsets - 1``
+    degrees of freedom. For details, see [1]_.
+
+    References
+    ----------
+    .. [1] Joe, Stephen. "Randomization of lattice rules for numerical multiple
+           integration." Journal of Computational and Applied Mathematics 31.2
+           (1990): 299-304.
+
+    Examples
+    --------
+    QMC quadrature is particularly useful for computing integrals in higher
+    dimensions; an example integrand is the probability density function
+    of a multivariate normal distribution.
+
+    >>> import numpy as np
+    >>> from scipy import stats
+    >>>
+    >>> dim = 8
+    >>> mean = np.zeros(dim)
+    >>> cov = np.eye(dim) / 2**8
+    >>>
+    >>> def func(*z):
+    ...     z = np.asarray(z).T
+    ...     return stats.multivariate_normal.pdf(z, mean, cov)
+
+    To compute the integral over the unit hypercube:
+
+    >>> from scipy.integrate import qmc_quad
+    >>> lb = np.zeros(dim)
+    >>> ub = np.ones(dim)
+    >>> ranges = np.asarray([lb, ub]).T
+    >>> rng = np.random.default_rng(1638083107694713882823079058616272161)
+    >>> qrng = stats.qmc.Halton(d=dim, seed=rng)
+    >>> n_points, n_offsets = 256, 8
+    >>> res = qmc_quad(func, ranges)
+    >>> res.integral, res.standard_error
+    (0.0001842819684617149, 1.3959190979748066e-07)
+
+    A two-sided, 99% confidence interval for the integral may be estimated
+    as:
+
+    >>> t = stats.t(df=n_offsets-1, loc=res.integral, scale=res.standard_error)
+    >>> t.interval(0.99)
+    (0.00018389017561108015, 0.00018461661169997918)
+
+    Indeed, the value reported by `scipy.stats.multivariate_normal.cdf` is
+    within this range.
+
+    >>> stats.multivariate_normal.cdf(ub, mean, cov, lower_limit=lb)
+    0.00018430867675187443  # may vary
+
+    """
+    args = _qmc_quad_iv(func, ranges, n_points, n_offsets, qrng, log, args)
+    func, ranges, n_points, n_offsets, qrng, rng, log, args, stats = args
+
+    dim = qrng.d
+
+    sample = qrng.random(n_points)  # shape is also (n_points, dim)
+    offsets = rng.random(size=(n_offsets, dim))
+    lb, ub = ranges.T
+    A = np.prod(ub - lb)
+    dA = A / n_points
+
+    # Typically, I'd use broadcasting and do this all in one go. I don't
+    # think that's the right way to go here. It's bad for memory, for
+    # large `n_points` and small `n_offsets` the speed impact is small (or
+    # negative), and `qmc.scale` is only for 2D arrays. Let's keep it simple.
+    estimates = np.zeros(n_offsets)
+    for i, offset in enumerate(offsets):
+        x = (sample + offset) % 1  # offset and wrap
```

Comment:
I am really concern about this part and believe that we are going to decrease the convergence rate.

If we would want to use an offset, it would need to be small enough to not move points away from their local hypercube. This is what we do with LHS for instance with small perturbations. On Sobol' it's also common to add a random shift (can be also constant like `1/(2n)` which works good), we do this (and something else) when we use `scramble`.

For the wrapping, this will create some imbalance as you are going to move some points from one part of the design to another. As we know, a single point can make a Sobol' sample go from a rate of `1/(n**3/2)` to `1/n`, this is going to do the same thing here.

Instead we really want to call R times `qrng.random`.

Since we require `qrng` to be a `QMCEngine`, we have access to `QMCEngine.rng`. From there we could use `QMCEngine.rng.__get_state__` to access the seed, then create a `SeedSequence` and spawn R generators that we can use to create R `qrng`.

@rkern is this how we should do it? I did not see something like `Generator.spawn`, so I am not sure what is the best thing to do here. I think we want to go with all that trouble in case we decide to parallelise the loop.

Reply:
This "pseudorandom shift mod 1" is most theoretically justified for lattice QRNGs, but according to [our doc's main RQMC reference](https://www.iro.umontreal.ca/~lecuyer/myftp/papers/rqmc-survey-2002.pdf) seems to work okay for others as well. But yes, using the `QMCEngine`-specific scrambling would probably be better.

When you do get to parallelizing, this is a simple, blessed way to implement `Generator` spawning until we lift that API up onto `Generator` itself:

```python
def rng_spawn(rng, n_children):
    bg = rng._bit_generator
    ss = bg._seed_seq
    child_rngs = [np.random.Generator(type(bg)(child_ss))
                  for child_ss in ss.spawn(n_children)]
    return child_rngs
```

Reply:
iIUC, instead of generating $R$ offsets and using those, you both suggest that I spawn $R$ Generators and from those create $R$ QMCEngines from which to greater the points.

@tupui can I just `reset` the original engine and swap out it's Generator attributes rather than instantiating many `QMCEngine`s that are supposed to be identical besides their generator? Shall I write this code in `QMCEngine` itself rather than in this function?

Reply:
> iIUC, instead of generating R offsets and using those, you both suggest that I spawn R Generators and from those create R QMCEngines from which to draw new points. Is that right?

Yes

> @tupui can I just `reset` the original engine and swap out it's Generator attributes rather than instantiating many `QMCEngine`s that are supposed to be identical besides their generator? Shall I write this code in `QMCEngine` itself rather than in this function?

You would need to call the init again at least for Sobol' because it does the scrambling there using the RNG. So if you swap the RNG, you need to re-initialize the class to update the matrix.

Reply:
I replaced this code in your script:
```python3
    # Sobol'
    engine = qmc.Sobol(d=case.dim, scramble=False)
    conv_res = conv_method(engine.random, case.func, ns, 1, case.ref)
    sample_sobol_rmse.append(conv_res)
```

with 
```python3
    # Sobol' + 
    engine = qmc.Sobol(d=case.dim, scramble=False)
    rng = np.random.default_rng()
    def myrandom(*args, **kwds):
        rand = engine.random(*args, **kwds)
        return (rand + rng.random(size=rand.shape[-1])) % 1
    conv_res = conv_method(myrandom, case.func, ns, 1, case.ref)
    sample_sobol_rmse.append(conv_res)
```
and it produces plots that don't look much different than if I use regular `Sobol` with `scramble=True`.

Maybe you can tell which is which : )
![image](https://user-images.githubusercontent.com/6570539/187743464-8345d8a4-263b-49b1-b5f9-fe79143cfa46.png)

![image](https://user-images.githubusercontent.com/6570539/187743659-742bfa33-49d2-4053-b0b0-389df9d80e0d.png)

Reply:
I would be surprised if this would hold for other integrands. But yes you might get better than unscrambled with some random shift (still surprised though it works that well due to the wrapping and not respecting subpace partitioning of the points).

But I would not expect this to perform better than proper scrambling as the scrambling technique we have was shown to be one, if not the best, method.

Reply:
Meh I'll ruin the fun. You _can_ tell from the scale. `Sobol` without scrambling gets the error down to $10^-2$.

![image](https://user-images.githubusercontent.com/6570539/187747299-35976fe2-db9f-4f92-9ca6-019ff6995f2c.png)

`Sobol` _with_ scrambling routinely gets the error down to $10^-7$ as shown (immediately) above.

Our modified versions - either with the offsets or sampling from multiple scrambled `Sobol`s with different `Generator`s - all seem to be better than unscrambled `Sobol`, but not as good as vanilla scrambled `Sobol`.

Reply:
> Our modified versions - either with the offsets or sampling from multiple scrambled `Sobol`s with different `Generator`s - all seem to be better than unscrambled `Sobol`, but not as good as vanilla scrambled `Sobol`.

I am not sure I understand exactly what you are saying. Did you try the following? Because this should give you the good rate. This is what I normally use.

```python
def _sampler_sobol_scrambled(ns):
    engine = qmc.Sobol(d=case.dim, scramble=True)
    return engine.random(ns)

conv_res = conv_method(_sampler_sobol_scrambled, case.func, ns, 99, case.ref)
```

Reply:
I can rephrase the statement you quoted, because I don't understand why that led you to suggest this code. (Changing from unscrambled Sobol' to scrambled Sobol' does not required definition a new function like `_sampler_sobol_scrambled(ns)`. I just required changed the example in the original script from `scrambled=False` to `scrambled=True`.)

> For the example problem 'Art 2', scrambled Sobol is the best, followed by all the other methods I tried. In last place is Sobol without scrambling. In higher dimensions, it doesn't seem to matter; all seem equally good on average.

Reply:
To be clear, the reason I am using this custom `_sampler_sobol_scrambled` is to recreate a new object for all 99 replicates.

The hierarchy you observed seems good.

Reply:
I am just seeing your code and I am not sure you are having replicates. If you are using my `conv_method`, you need to change the replicates which here you set to 1.

Reply:
I didn't know that there were supposed to be replicates, I guess to average the results out? That would be nice. I can take a look.

Reply:
I guess you wanted me to use the [second link](https://gist.github.com/tupui/fb6e219b1dd2316b7498ebce231bfff5). I was using [the first](https://github.com/scipy/scipy/blob/main/doc/source/tutorial/stats/plots/qmc_plot_conv_mc_sobol.py).

Reply:
For anyone, we had a chat and I gave another script

<details>
<summary>Details</summary>

```python
"""Integration convergence using Sobol' sequence.

MIT License

Copyright (c) 2020 Pamphile Tupui ROY
"""
import os
from collections import namedtuple

import numpy as np
from scipy.stats import qmc

import matplotlib.pyplot as plt
from matplotlib.container import ErrorbarContainer
from matplotlib.legend_handler import HandlerErrorbar

# Parameters
dim = 30
n_conv = 99

MIN_POWER = 4
POWER_MAX = 14
ns_gen = 2 ** np.arange(MIN_POWER, POWER_MAX + 1)  # 13

path = f'sobol_convergence/centre'
os.makedirs(path, exist_ok=True)


# Functions definitions
_exp1 = 1 - np.exp(1)


def art_0(sample):
    # dim 2, true value 0
    return np.prod(sample - 0.5, axis=1)


def art_1d(sample):
    # dim 1, true value 1
    return sample * np.exp(sample)


def art_1(sample):
    # dim 5, true value 0
    return np.sum(np.exp(sample) + _exp1, axis=1)


def art_2(sample):
    # dim 5, true value 5/3 + 5*(5 - 1)/4
    return np.sum(sample, axis=1) ** 2


def art_3(sample):
    # dim 3, true value 0
    return np.prod(np.exp(sample) + _exp1, axis=1)


def type_a(sample, dim=30):
    # true value 1
    a = np.arange(1, dim + 1)
    f = 1.
    for i in range(dim):
        f *= (abs(4. * sample[:, i] - 2) + a[i]) / (1. + a[i])
    return f


def type_b(sample, dim=30):
    # true value 1
    f = 1.
    for d in range(1, dim + 1):
        f *= (d - sample[:, d - 1]) / (d - 0.5)
    return f


def type_b_f2(sample, dim=30):
    # true value 1
    f = 1.
    inv = 1. / dim
    for d in range(dim):
        f *= (1. + 1. / dim) * sample[:, d] ** inv
    return f


def type_c(sample, dim=10):
    # true value 1
    f = 2 ** dim * np.prod(sample, axis=1)
    return f


def additive(sample):
    # true value 1/2*dim
    return np.sum(sample, axis=1)


def pi(x):
    # true value np.pi
    radius = (x[:, 0]**2 + x[:, 1]**2) ** .5
    in_circle = len(np.where(radius <= 1)[0])
    return 4 * in_circle / x.shape[0]


functions = namedtuple('functions', ['name', 'func', 'dim', 'ref'])

benchmark = [
    functions('Art 2D', art_0, 2, 0),
    functions('Art 1D', art_1d, 1, 1),
    functions('Art 1', art_1, 5, 0),
    functions('Art 2', art_2, 5, 5 / 3 + 5 * (5 - 1) / 4),
    functions('Art 3', art_3, 3, 0),
    functions('Type A', type_a, 30, 1),
    functions('Type B f3', type_b, 30, 1),
    functions('Type B f2', type_b_f2, 30, 1),
    functions('Type C', type_c, 10, 1),
    functions('Additive', additive, 30, 1/2*30)
]

benchmark = [func for func in benchmark if func.dim == dim]


def conv_method(samples, func, n_samples, n_conv, ref):
    evals = [np.sum(func(sample)) / n_samples for sample in samples]
    squared_errors = (ref - np.array(evals)) ** 2
    rmse = (np.sum(squared_errors) / n_conv) ** 0.5

    return rmse, 0.  # can add another metric


# Analysis

def sampler_mc(ns, dim):
    sampler_ = lambda x: np.random.random((x, dim))
    samples_ = [sampler_(ns) for _ in range(n_conv)]
    return np.array(samples_)


def sampler_sobol(ns, dim):
    def _sampler_sobol(ns):
        engine = qmc.Sobol(d=dim, scramble=False, bits=64)
        return engine.random(ns)

    samples_ = [_sampler_sobol(ns) for _ in range(n_conv)]
    return np.array(samples_)


def sampler_sobol_col_shift(ns, dim):
    def _sampler_sobol(ns):
        engine = qmc.Sobol(d=dim, scramble=False, bits=64)
        sample = engine.random(ns)
        return sample[:, np.random.choice(dim, dim, replace=False)]

    samples_ = [_sampler_sobol(ns) for _ in range(n_conv)]
    return np.array(samples_)


def sampler_sobol_centre(ns, dim):
    def _sampler_sobol(ns):
        engine = qmc.Sobol(d=dim, scramble=False, bits=64)
        sample = engine.random(ns)
        return sample + 1/(2*ns)

    samples_ = [_sampler_sobol(ns) for _ in range(n_conv)]
    return np.array(samples_)


def sampler_sobol_scrambled_64bits(ns, dim):
    def _sampler_sobol(ns):
        engine = qmc.Sobol(d=dim, scramble=True, bits=64)
        return engine.random(ns)

    samples_ = [_sampler_sobol(ns) for _ in range(n_conv)]
    return np.array(samples_)


samplers = {
    "MC": [sampler_mc, []],
    "Sobol' unscrambled": [sampler_sobol, []],
    "Sobol' scrambled": [sampler_sobol_scrambled_64bits, []],
    # "Sobol' column shift": [sampler_sobol_col_shift, []],
    "Sobol' centre": [sampler_sobol_centre, []],
}


for ns in ns_gen:
    print(f'-> ns={ns}')

    for sampler_ in samplers:
        sample_ = samplers[sampler_][0](ns, dim)

        conv_res = []
        for case in benchmark:
            conv_res_ = conv_method(sample_, case.func, ns, n_conv, case.ref)
            conv_res.append(conv_res_)

        samplers[sampler_][1].append(conv_res)


for sampler_ in samplers:
    samplers[sampler_][1] = np.array(samplers[sampler_][1])


# Plot
i = -1
for case in benchmark:
    i += 1

    func = case.name
    fig, ax = plt.subplots()

    # ratio_1 = sample_sobol_0_rmse[:, i, 0][0] / ns_gen[0] ** (-2/2)
    # ratio_1 = sample_sobol_scramble_rmse[:, i, 0][0] / ns_gen[0] ** (-2 / 2)
    # ratio_2 = sample_sobol_scramble_0_rmse[:, i, 0][0] / (np.log2(ns_gen[0]) * ns_gen[0] ** (-3 / 2))
    # ratio_3 = sample_sobol_scramble_0_rmse[:, i, 0][0] / (ns_gen[0] ** (-3 / 2))

    # ax.plot(ns_gen, ns_gen ** (-1 / 2), ls='-', c='k')
    # ax.plot(ns_gen, ns_gen ** (-2/2) * ratio_1, ls='-', c='k')
    # ax.plot(ns_gen, np.log2(ns_gen) * ns_gen ** (-3 / 2) * ratio_2, ls='-.')
    # ax.plot(ns_gen, ns_gen ** (-3 / 2) * ratio_3, ls='-', c='k')

    for sampler_ in samplers:
        ax.plot(
            ns_gen, np.log2(samplers[sampler_][1][:, i, 0]),
            label=sampler_
        )

    ax.set_xlabel(r'$N_s$')

    ax.set_xscale('log')
    # ax.set_yscale('log')

    ax.set_xticks(ns_gen)
    ax.set_xticklabels([fr'$2^{{{ns}}}$' for ns in np.arange(MIN_POWER, POWER_MAX + 1)])
    ax.set_ylabel(r'$\log (\epsilon (\mu))$')
    fig.legend(labelspacing=0.7, bbox_to_anchor=(0.5, 0.45),
               handler_map={ErrorbarContainer: HandlerErrorbar(xerr_size=0.7)})
    fig.tight_layout()
    # plt.show()
    fig.savefig(os.path.join(path, f'sobol_conv_integration_mean_{func}.pdf'),
                transparent=True, bbox_inches='tight')

```
</details>

Reply:
Here are the plots for the 30D problems.
If $n$ is the total number of function evaluation points and $R$ is the number of independent estimates of the integral (each from evaluating the function at $n/R$ points), then:
* "Sobol' unscrambled, single offset" generates $n$ points from unscrambled Sobol' and applies $R=1$ pseudorandom offset
* "Sobol' unscrambled, single offset" generates $n/R$ points from unscrambled Sobol' and applies $R=4$ pseudorandom offsets to generate a total of $n$ function evaluation points
*  "Sobol' multi-scrambled" generates $n/R$ points from each of $R=4$ scrambled Sobol' sequences, each with a different underlying PRNG.

![image](https://user-images.githubusercontent.com/6570539/187766260-a8bac41d-6de4-48b8-97f6-d7544f48589b.png)

![image](https://user-images.githubusercontent.com/6570539/187767088-002ff350-2db2-4f27-88b9-271e31abda66.png)

![image](https://user-images.githubusercontent.com/6570539/187767138-9dd8e133-59ac-4bf7-a706-b5513d19fcdd.png)

![image](https://user-images.githubusercontent.com/6570539/187767202-b38fd35e-4eb9-459d-b764-daac08378974.png)

It does look like the offset methods tend to have higher error than Sobol' multi-scrambled, and they even seem to lose the convergence rate for Tybe Bf3.

Sobol' multi-scrambled is always a little worse than Sobol' scrambled, but only by a constant offset (on the log plot). Let's go with that (as you suggested in the beginning). We'll advise users that they can choose $R=1$ if they want the best overall integral estimate with a given number of evaluation points, but we will not be able to return an error estimate. Users can increase $R > 1$ for a (typically) less accurate integral estimate but (typically) more accurate error estimates.

Reply:
Great thanks for doing that Matt :rocket: Glad that it does confirm the intuition/explanation I had.

For completeness, Art is currently investigating using the median instead of the mean to aggregate the samples and apparently it provides better results. But he does not yet have something to get back a confidence interval. https://arxiv.org/abs/2208.05078

Reply:
I'll resolve this part. We can consider that in a follow-up. (But why not just bootstrap the CI on the median?)